### ISSUE TITLE
pygmt.Figure.meca offset not working

### ISSUE THREAD

**Issue by rdcorona** [GenericMappingTools/pygmt#1129]
I am using pygmt to graph focal mechanisms with magnitude scale.
However, when I try to give it an offset so that they don't pile up the function doesn't seem to work.

I'm using :

```python
spec = dict (strike=strikes[0], dip=dips[0], rake=rakes[0], magnitude=magnitude [0])

fig.meca (spec, scale=f ".{0.002 * (2 ** magnitude[0])}c", longitude = lons [0], latitude = lats [0], depth = 25.0, convention = "aki", plot_longitude = lons [0], plot_latitude = lats[0] -10, offset=True) 
```
where all the objects are lists with real (float) values

**System information**
```
# Name                    Version                   Build  Channel
pygmt                     0.3.1              pyhd8ed1ab_0    conda-forge

```

Do I make something wrong?

**Comment by core-man** [comment]
I can re-produce this potential bug using [Focal mechanisms](https://www.pygmt.org/dev/gallery/seismology/meca.html#sphx-glr-gallery-seismology-meca-py) gallery example.

```Python
import pygmt
fig = pygmt.Figure()
fig.coast(region=[-125, -122, 47, 49], projection="M6c", land="grey", water="lightblue", shorelines=True, frame="a")
focal_mechanism = dict(strike=330, dip=30, rake=90, magnitude=3)
fig.meca(focal_mechanism, scale="1c", longitude=-124.3, latitude=48.1, depth=12.0, plot_longitude=-124.3, plot_latitude=47.5, offset=True)
fig.show()
```
![meca](https://user-images.githubusercontent.com/50591376/112717373-32942c00-8f27-11eb-934e-d9924d1e983e.png)

**Comment by seisman** [comment]
I can reproduce the issue, and now know why it fails. Unfortunately, there is no workaround now.

**Comment by rdcorona** [comment]
Hello seisman.
So, we'll have to wait for the 0.4.0 release.🤷🏻‍♂️

Thanks for the reply

**Comment by seisman** [comment]
Currently, offsetting beachballs only works for files. 

I'll see what I can do to fix it.

**Comment by rdcorona** [comment]
Ok, meanwhile I'll try it with a dataFrame and an ASCII file 

Thanks again

**Comment by hazardgoat** [comment]
> Ok, meanwhile I'll try it with a dataFrame and an ASCII file

I know this is an old thread, but I too am unable to get the focal mechanism offset parameter to work. Were you ever able to find a way to get it to offset?

**Comment by weiji14** [comment]
> I can reproduce the issue, and now know why it fails. Unfortunately, there is no workaround now.

Still a problem on PyGMT v0.5.0 unfortunately. Could you elaborate more on why it fails? I tried debugging this while handling #1611, and couldn't see any obvious issue on the Python side, e.g. the `plot_longitude` and `plot_latitude` variables seem to be appended to `spec` ok here:

https://github.com/GenericMappingTools/pygmt/blob/9e8a7bd81188f87637a7611239db30283f641b39/pygmt/src/meca.py#L330-L337

Only thing I can think of is that this is a problem with the `spec` array to GMT virtualfile convention at https://github.com/GenericMappingTools/pygmt/blob/9e8a7bd81188f87637a7611239db30283f641b39/pygmt/src/meca.py#L454 so should this be marked as an upstream GMT API bug then?

**Comment by seisman** [comment]
> Ah I see, maybe we could use `lib.virtualfile_from_data` as with #949?

Yes, ideally we should use `lib.virtualfile_from_data`.

> Just to be clear, do you mean that `plot_longitude`, `plot_latitude` and `event_name` is to be passed in as 3 columns or 1 column using `GMT_Put_Strings`?

We can only call `GMT_Put_Strings` once, so we have to combine the three parameters into one string vector and pass it to `GMT_Put_Strings`. Actually, we already did it in the `virtualfile_from_vector` function. The `virtualfile_from_vector` function doesn't work for `meca`, because plot_longitude and plot_latitude are numeric vectors but must be passed as strings.

https://github.com/GenericMappingTools/pygmt/blob/4b80f3c269f39917c8bb7ee96c5a253279877024/pygmt/clib/session.py#L1159-L1172.

**Comment by seisman** [comment]
I don't know what happened to GitHub. 

My reply to your comment https://github.com/GenericMappingTools/pygmt/issues/1129#issuecomment-962574090 is displayed before your comment at here: https://github.com/GenericMappingTools/pygmt/issues/1129#issuecomment-962576217

**Comment by weiji14** [comment]
> I don't know what happened to GitHub.
> 
> My reply to your comment [#1129 (comment)](https://github.com/GenericMappingTools/pygmt/issues/1129#issuecomment-962574090) is displayed before your comment at here: [#1129 (comment)](https://github.com/GenericMappingTools/pygmt/issues/1129#issuecomment-962576217)

Yeah I see the same thing, must be time going backwards :laughing:

> > Ah I see, maybe we could use `lib.virtualfile_from_data` as with #949?
> 
> Yes, ideally we should use `lib.virtualfile_from_data`.
> 
> > Just to be clear, do you mean that `plot_longitude`, `plot_latitude` and `event_name` is to be passed in as 3 columns or 1 column using `GMT_Put_Strings`?
> 
> We can only call `GMT_Put_Strings` once, so we have to combine the three parameters into one string vector and pass it to `GMT_Put_Strings`. Actually, we already did it in the `virtualfile_from_vector` function. The `virtualfile_from_vector` function doesn't work for `meca`, because plot_longitude and plot_latitude are numeric vectors but must be passed as strings.

Ok, I think I got a clear idea on what to do now. We'll need to convert the `plot_longitude` and `plot_latitude` vectors from numerical arrays to string arrays, and then `lib.virtualfile_from_vectors` (via `lib.virtualfile_from_data`) should do the concatenation magic (joining `event_name` if it is present) and call `GMT_Put_Strings` only once on the concatenated string column.

Probably do 2 separate PRs. 1) Refactor from using `lib.virtualfile_from_matrix` to `lib.virtualfile_from_data` (which will enable other input data structures like `xarray.Dataset` arrays) and 2) Do the actual bugfix so that `offset` works.

**Comment by seisman** [comment]
> Ok, I think I got a clear idea on what to do now. We'll need to convert the `plot_longitude` and `plot_latitude` vectors from numerical arrays to string arrays, and then `lib.virtualfile_from_vectors` (via `lib.virtualfile_from_data`) should do the concatenation magic (joining `event_name` if it is present) and call `GMT_Put_Strings` only once on the concatenated string column.
> 
> Probably do 2 separate PRs. 1) Refactor from using `lib.virtualfile_from_matrix` to `lib.virtualfile_from_data` (which will enable other input data structures like `xarray.Dataset` arrays) and 2) Do the actual bugfix so that `offset` works.

Yes!

**Comment by seisman** [comment]
The input parameters of the `meca` module are a little complicated. I will take the Aki & Richard convention (**-Sa**) as an example. The input parameters are:
```
longitude latitude depth strike dip rake magnitude [plot_longitude plot_latitude] [event_name]
```
As you can see, plot_longitude & plot_latitude and event_name are optional. Thus users can give input parameters in many different ways:

1. don't give plot longitude&latitude and event_name
2. give plot longitude&latitude but not event_name
3. give event_name but not plot longitude&latitude

Due to these complications, the first few columns (7 columns for the **-Sa** case) are read as numeric values (`S->data in the GMT source code)`, thus data passed as vectors (`virtualfile_from_vector`) or a matrix (`virtualfile_from_matrix`) can be read correctly. However, because GMT can't know exactly the number of trailing columns and the event_name can contain whitespace, all the trailing columns are read as trailing text (`S->text` in the GMT source code). 

See https://github.com/GenericMappingTools/gmt/blob/3e68b8de02774f2eda0068ea356586d61770f070/src/seis/psmeca.c#L775  and https://github.com/GenericMappingTools/gmt/blob/3e68b8de02774f2eda0068ea356586d61770f070/src/seis/psmeca.c#L824 if you want to read the GMT source code.

Due to the above reasons, we have to pass the trailing columns (plot longitude&latitude and event_name) using `GMT_Put_Strings`, not `GMT_Put_Matrix` or `GMT_Put_Vectors`. Actually, we once had similar issues with the `text` function (related to https://github.com/GenericMappingTools/pygmt/issues/483, https://github.com/GenericMappingTools/pygmt/pull/520).

Thus, what we need to do is:

- Replace `lib.virtualfile_from_matrix` with `lib.virtualfile_from_vector`
- Build a string list/vector for `plot_longitude plot_latitude event_name` and pass it to `lib.virtual_from_vector`.

**Comment by weiji14** [comment]
> Thus, what we need to do is:
> 
> * Replace `lib.virtualfile_from_matrix` with `lib.virtualfile_from_vector`
> * Build a string list/vector for `plot_longitude plot_latitude event_name` and pass it to `lib.virtual_from_vector`.

Ah I see, maybe we could use `lib.virtualfile_from_data` as with #949? Since it basically passes things to `lib.virtualfile_from_vectors` anyways:

https://github.com/GenericMappingTools/pygmt/blob/4b80f3c269f39917c8bb7ee96c5a253279877024/pygmt/clib/session.py#L1427-L1431

though it sounds like we'll need to be careful and ensure that it doesn't get into this case:

https://github.com/GenericMappingTools/pygmt/blob/4b80f3c269f39917c8bb7ee96c5a253279877024/pygmt/clib/session.py#L1454-L1458

so some very careful refactoring and unit test writing to do!

**Comment by weiji14** [comment]
> > Ok, I think I got a clear idea on what to do now. We'll need to convert the `plot_longitude` and `plot_latitude` vectors from numerical arrays to string arrays, and then `lib.virtualfile_from_vectors` (via `lib.virtualfile_from_data`) should do the concatenation magic (joining `event_name` if it is present) and call `GMT_Put_Strings` only once on the concatenated string column.
> > Probably do 2 separate PRs. 1) Refactor from using `lib.virtualfile_from_matrix` to `lib.virtualfile_from_data` (which will enable other input data structures like `xarray.Dataset` arrays) and 2) Do the actual bugfix so that `offset` works.
> 
> Yes!

On third thought, maybe 3 separate PRs (because `pygmt.meca` is a beast to refactor...):

1. Refactor `meca` to use `virtualfile_from_data` (doing at #1613)
2. Handle text labels and angle/font/justify (c.f. #1466)
3. Fix this `offset` parameter bug (#1129).

**Comment by weiji14** [comment]
Apparently the offset bug wasn't properly fixed in #1784 for PyGMT v0.7.0, but there's a workaround at https://github.com/GenericMappingTools/pygmt/issues/2016#issuecomment-1326667264 (basically use `str` types for `plot_longitude` and `plot_latitude` inputs), and we're looking at fixing things properly in #2202 for PyGMT v0.8.0.
